File: inJava/app/src/main/java/com/barrygithub/rickandmortyapp/data/RepositoryImpl.java
package com.barrygithub.rickandmortyapp.data;

import com.barrygithub.rickandmortyapp.data.LocalDatasource.MyDAO;
import com.barrygithub.rickandmortyapp.data.LocalDatasource.entities.Character;
import com.barrygithub.rickandmortyapp.data.LocalDatasource.entities.EntityDb;
import com.barrygithub.rickandmortyapp.data.LocalDatasource.entities.MetaData;
import com.barrygithub.rickandmortyapp.data.RemoteDatasource.ApiClient;
import com.barrygithub.rickandmortyapp.graphql.Queries;

import org.json.JSONObject;

import java.util.List;

import io.reactivex.rxjava3.core.Single;

/**
 * Project RickAndMortyApp
 * Created by Barry Zea H. on 3/04/23.
 * Copyright (c) Barry Zea H. All rights reserved.
 **/

public class RepositoryImpl  implements Repository{
    ApiClient apiClient;
    MyDAO db;
    public RepositoryImpl(ApiClient apiClient, MyDAO db){
        this.apiClient=apiClient;
        this.db= db;
    }

    @Override
    public Single<String> getDataFromApiGraphql(int page) {
        JSONObject paramObject = new JSONObject();
        try{
        paramObject.put("query", Queries.getAllCharactersQuery(page));}
        catch(Exception e){}
        return apiClient.getDataFromApiGraphql(paramObject.toString());

   }

    @Override
    public Single<EntityDb> getDataFromDb() {
        return  db.getDataFromDatabase();

    }

    @Override
    public void saveMetadataInDb(MetaData metadata) {
        //seteamos manualmente el id aunque esté como autogenrado en room
        //lo hacemos para que no vuelva a insertar un nuevo registro y reemplace el mismo que ya tenemos la primera vez
        //y que el método de insertar tiene como anotación OnConflictStrategy.REPLACE
        //Otra opción es  guardar solamente la lista de personajes (saveCharacters) obviando metadata solo para la API
        metadata.setIdMetadata(1);
        db.saveMetadata(metadata);
    }

    @Override
    public void saveCharactersInDb(List<Character> characters) {
      db.saveCharacters(characters);
    }


}

File: inKotlin/app/src/main/java/com/barrygithub/rickandmortyapp/data/remoteDatasource/RepositoryImpl.kt
package com.barrygithub.rickandmortyapp.data.remoteDatasource

import com.barrygithub.rickandmortyapp.data.localDatasource.Character
import com.barrygithub.rickandmortyapp.data.localDatasource.MetaData
import com.barrygithub.rickandmortyapp.data.localDatasource.MyDAO
import com.barrygithub.rickandmortyapp.graphql.getCharacterQuery
import io.reactivex.rxjava3.core.Single
import org.json.JSONObject

/**
 * Project RickAndMortyApp
 * Created by Barry Zea H. on 14/04/23.
 * Copyright (c) Barry Zea H. All rights reserved.
 *
 **/
class RepositoryImpl(private val api:ApiClient, private val db:MyDAO) : Repository{

    //gracias a la query de graphQl nos deveolverá una lista de capítulos de cada personaje
    //y no una lista de enlaces como en la rest API, así que ya no necesitaremos llamar a cada enlace de episodio
    //para obtener sus campos

    override fun getDataFromGraphql(page: Int): Single<String> {
        //Convertimos nuestra query en un objeto tipo json para la peticion a la api graphql
        val paramObject = JSONObject()
        paramObject.put("query", getCharacterQuery(page))
        return api.getDataFromApiGraphQl(paramObject.toString())

    }

    override fun saveMetadataInDb(metadata: MetaData) {
        //seteamos manualmente el id aunque esté como autogenrado en room
        //lo hacemos para que no vuelva a insertar un nuevo registro y reemplace el mismo que ya tenemos la primera vez
        //y que el método de insertar tiene como anotación OnConflictStrategy.REPLACE
        //Otra opción es  guardar solamente la lista de personajes (saveCharacters) obviando metadata solo para la API

        metadata.idMetadata =1
        db.saveMetadata(metadata);
    }

    override fun saveCharactersInDb(characters: List<Character>) {
        db.saveCharacters(characters)
    }

    override fun getDataFromDb() = db.getDataFromDatabase()

}